### DISCUSSION TITLE
How to create and launch UI tests on NetBeans platform and NetBeans 23 IDE

### DISCUSSION THREAD

**Discussion by Sova777** [apache/netbeans#7867] [Tips, Tricks and Workarounds]
### Create UI test for NetBeans platform

- Press New Project... button on toolbar and select NetBeans Application project type. Create project with default settings.
<img width="539" alt="nb_ui_testing_01" src="https://github.com/user-attachments/assets/3938f650-8940-4eba-9fa7-bf89008e94b5">

- Select *-parent project and press Build Project toolbar button. Open ApplicationTest.java file in *-app project. Remove comments in testApplication() method.

<img width="496" alt="nb_ui_testing_02" src="https://github.com/user-attachments/assets/e89bd1d1-f0c0-475a-9270-8974e77d92a3">

- Click on lamp icon and add JellyTools library

<img width="607" alt="nb_ui_testing_03" src="https://github.com/user-attachments/assets/50b5886d-31da-40c6-9e6e-48f724f2481a">
<img width="608" alt="nb_ui_testing_04" src="https://github.com/user-attachments/assets/82da140e-75a2-47b0-99bf-50957d61bd7d">

- Now we can use library

<img width="521" alt="nb_ui_testing_05" src="https://github.com/user-attachments/assets/af789417-f114-4233-9728-1fd33882b8d9">
<img width="695" alt="nb_ui_testing_06" src="https://github.com/user-attachments/assets/8c2ab015-02a6-4cec-9df0-460bfb5c3b5d">

- Replace gui(false). on gui(true).

<img width="521" alt="nb_ui_testing_07" src="https://github.com/user-attachments/assets/8af799fc-b7b9-4c0b-8ea0-5ca800bb0976">

- Launch test. Ok. No errors.

<img width="503" alt="nb_ui_testing_08" src="https://github.com/user-attachments/assets/a65d84e1-b4eb-476a-8a06-cf71c39cb5ca">

### Create UI test for NetBeans IDE

- Open pom.xml file from *-app project. Replace all.clusters value on path to NetBeans IDE.

<img width="634" alt="nb_ui_testing_09" src="https://github.com/user-attachments/assets/d158a581-b3a0-447d-ace9-930d7c68aa00">

- Launch test. Test has been completed, but NetBeans IDE started with errors.

<img width="1004" alt="nb_ui_testing_10" src="https://github.com/user-attachments/assets/a92a8170-ea34-4213-abb7-baf71b8e2f39">

- In order to fix errors we need to add Java flags from original NetBeans. Options can be found in IDE Log tab (View > IDE Log menu item). You can restart NetBeans in order to see more short log.

<img width="674" alt="nb_ui_testing_11" src="https://github.com/user-attachments/assets/efe9f223-e133-4b36-8db2-945ada48b8b4">

- Add Java flags to pom.xml

<img width="661" alt="nb_ui_testing_12" src="https://github.com/user-attachments/assets/fd60e299-cf98-434c-bdaf-c8c6814ae463">

- Unfortunately test fails. And we need to replace failOnMessage(Level.WARNING). on failOnMessage(Level.SEVERE). in ApplicationTest.java too.

<img width="725" alt="nb_ui_testing_13" src="https://github.com/user-attachments/assets/dfccce88-802f-4f6c-a10b-a22bf3891d10">

- Add some code to test and check again. Test works nicely. No errors.

<img width="878" alt="nb_ui_testing_14" src="https://github.com/user-attachments/assets/8cea0e08-0bb7-4438-8c13-af6faf599376">

**Comment by Chris2011** [comment]
hey @Sova777 thank you very much for this cool tutorial. There is an official Wiki where it should be better put this: https://netbeans.apache.org/wiki/main/wiki/ If you want. It is like the old wiki how to do thngs :). Great work.

**Comment by Sova777** [comment]
Example of real test. Create Java project, open Project Properties window, switch off Compile on Save checkbox:
```java
        final String SAMPLE_PROJECT_NAME = "uitest_" + System.currentTimeMillis();

        NewProjectWizardOperator npwo = NewProjectWizardOperator.invoke();
        // "Standard"
        String standardLabel = "Java with Maven";
        npwo.selectCategory(standardLabel);
        // "Java Application"
        String javaApplicationLabel = "Java Application";
        npwo.selectProject(javaApplicationLabel);
        npwo.next();
        NewJavaProjectNameLocationStepOperator npnlso = new NewJavaProjectNameLocationStepOperator();
        npnlso.txtProjectName().setText(SAMPLE_PROJECT_NAME);
        npnlso.txtProjectLocation().setText(System.getProperty("netbeans.user"));
        npnlso.btFinish().pushNoBlock();
        npnlso.getTimeouts().setTimeout("ComponentOperator.WaitStateTimeout", 120000);
        npnlso.waitClosed();
        // wait project appear in projects view
        new ProjectsTabOperator().getProjectRootNode(SAMPLE_PROJECT_NAME);

        //disable the compile on save:
        ProjectsTabOperator.invoke().getProjectRootNode(SAMPLE_PROJECT_NAME).properties();
        // "Project Properties"
        String projectPropertiesTitle = "Project Properties";
        NbDialogOperator propertiesDialogOper = new NbDialogOperator(projectPropertiesTitle);
        // select "Compile" category
        String buildCategoryTitle = "Build";
        String compileCategoryTitle = "Compile";
        new Node(new Node(new JTreeOperator(propertiesDialogOper), buildCategoryTitle), compileCategoryTitle).select();
        // actually disable the quick run:
        String compileOnSaveLabel = "Compile on Save";
        JCheckBox cb = JCheckBoxOperator.waitJCheckBox((Container) propertiesDialogOper.getSource(), compileOnSaveLabel, true, true);
        if (cb.isSelected()) {
            cb.doClick();
        }
        // confirm properties dialog
        propertiesDialogOper.ok();
```

**Comment by Sova777** [comment]
And video tutorial: https://www.youtube.com/watch?v=qPFJk6VqD3U

**Reply by Chris2011** [reply]
@Sova777 Great work, thx :). Does this help you? @ebresie
